packages/wrangler/src/pages/validate.tsx

```diff
@@ -0,0 +1,127 @@
+import { readdir, stat } from "node:fs/promises";
+import { join, relative, resolve, sep } from "node:path";
+import { getType } from "mime";
+import { Minimatch } from "minimatch";
+import prettyBytes from "pretty-bytes";
+import { FatalError } from "../errors";
+import { MAX_ASSET_COUNT, MAX_ASSET_SIZE } from "./constants";
+import { hashFile } from "./hash";
+
+import type {
+	CommonYargsArgv,
+	StrictYargsOptionsToInterface,
+} from "../yargs-types";
+
+type UploadArgs = StrictYargsOptionsToInterface<typeof Options>;
+
+export function Options(yargs: CommonYargsArgv) {
+	return yargs.positional("directory", {
+		type: "string",
+		demandOption: true,
+		description: "The directory of static files to upload",
+	});
+}
+
+export const Handler = async ({ directory }: UploadArgs) => {
+	if (!directory) {
+		throw new FatalError("Must specify a directory.", 1);
+	}
+
+	await validate({
+		directory,
+	});
+};
+
+export type FileContainer = {
+	path: string;
+	contentType: string;
+	sizeInBytes: number;
+	hash: string;
+};
+
+export const validate = async (args: {
+	directory: string;
+}): Promise<Map<string, FileContainer>> => {
+	const IGNORE_LIST = [
+		"_worker.js",
+		"_redirects",
+		"_headers",
+		"_routes.json",
+		"functions",
+		"**/.DS_Store",
+		"**/node_modules",
+		"**/.git",
+	].map((pattern) => new Minimatch(pattern));
+
+	const directory = resolve(args.directory);
+
+	// TODO(future): Use this to more efficiently load files in and speed up uploading
+	// Limit memory to 1 GB unless more is specified
+	// let maxMemory = 1_000_000_000;
+	// if (process.env.NODE_OPTIONS && (process.env.NODE_OPTIONS.includes('--max-old-space-size=') || process.env.NODE_OPTIONS.includes('--max_old_space_size='))) {
+	// 	const parsed = parser(process.env.NODE_OPTIONS);
+	// 	maxMemory = (parsed['max-old-space-size'] ? parsed['max-old-space-size'] : parsed['max_old_space_size']) * 1000 * 1000; // Turn MB into bytes
+	// }
+
+	const walk = async (
+		dir: string,
+		fileMap: Map<string, FileContainer> = new Map(),
+		startingDir: string = dir
+	) => {
+		const files = await readdir(dir);
+
+		await Promise.all(
```

Comment:
If there's 10,000k files in a directory, will this blow up?

Reply:
It's prooooobably worth pulling in p-queue

Reply:
This PR is just a refactor, and I'd rather keep it to just a copy-paste job right now rather than changing anything about how this actually works. Would totally be fine with creating a new issue to follow-up on improving this though!

Reply:
Yeahhh I realized that _after_ I made the comment

Reply:
Agreed - probably worth improving this code, but that's out of scope of this PR